### PR TITLE
ovl_Dm_Tsg: OK

### PR DESCRIPTION
**Pull request by sonicdcer**
image

### CODE REVIEW

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 1]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Tsg/DmTsg_Destroy.s")
+void DmTsg_Update(Actor* thisx, PlayState* play) {
+    DmTsg* this = THIS;
+    s32 cueChannel;
+    s16 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Tsg/DmTsg_Update.s")
+    for (i = 0; i < 100; i++) {
+        this->unk_5F8[i] = (Math_SinS((i & 0xF) * this->unk_854) * 60.0f) + 120.0f;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:55]: Would this match?
```suggestion
        this->unk_5F8[i] = (Math_SinS((i % 16) * this->unk_854) * 60.0f) + 120.0f;
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h [hunk 2]
```diff
@@ -5,9 +5,18 @@
 
 struct DmTsg;
 
+typedef void (*DmTsgActionFunc)(struct DmTsg*, PlayState*);
+
 typedef struct DmTsg {
     /* 0x000 */ Actor actor;
-    /* 0x144 */ char unk_144[0x7DC];
+    /* 0x144 */ DmTsgActionFunc ActionFunc;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h:12]: This doesn't seem to be used, why do you think it is an actionfunc?

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h [hunk 3]
```diff
@@ -5,9 +5,18 @@
...
+    /* 0x148 */ Vec3f unk_148[100];
+    /* 0x5F8 */ f32 unk_5F8[100];
+    /* 0x788 */ f32 unk_788;
+    /* 0x78C */ s16 unk_78C[100];
+    /* 0x854 */ s16 unk_854;
+    /* 0x856 */ s16 unk_856[100];
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h:18]: Seems like all of those arrays should always have the same size by looking at the code.
Could you make a define for the size (like `#define DMTSG_UNK_SIZE 100`) and use it everywhere?
```suggestion
    /* 0x148 */ Vec3f unk_148[DMTSG_UNK_SIZE];
    /* 0x5F8 */ f32 unk_5F8[DMTSG_UNK_SIZE];
    /* 0x788 */ f32 unk_788;
    /* 0x78C */ s16 unk_78C[DMTSG_UNK_SIZE];
    /* 0x854 */ s16 unk_854;
    /* 0x856 */ s16 unk_856[DMTSG_UNK_SIZE];
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 4]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
+        this->unk_148[i].x = Math_CosS(this->unk_78C[i]) * this->unk_5F8[i];
+        this->unk_148[i].y = this->unk_788 + (i * 16);
+        this->unk_148[i].z = Math_SinS(this->unk_78C[i]) * this->unk_5F8[i];
+        this->unk_78C[i] += this->unk_856[i];
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Tsg/DmTsg_Draw.s")
+    this->unk_91E = 0;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:62]: Seems to be used as a boolean
```suggestion
    this->unk_91E = false;
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 5]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
+    this->unk_854 += 40;
+
+    if (Cutscene_IsCueInChannel(play, CS_CMD_ACTOR_CUE_514)) {
+        cueChannel = Cutscene_GetCueChannel(play, CS_CMD_ACTOR_CUE_514);
+        if (play->csCtx.actorCues[cueChannel]->id != 1) {
+            this->unk_91E = 1;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:68]:
```suggestion
            this->unk_91E = true;
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 6]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
+    }
+}
+
+void DmTsg_Draw(Actor* thisx, PlayState* play2) {
+    PlayState* play = play2;
+    DmTsg* this = THIS;
+    s32 i;
+    u32 j;
+
+    if (this->unk_91E != 0) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:80]:
```suggestion
    if (this->unk_91E) {
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c (hunk 1 above)
Comment (sonicdcer) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:55]: It doesn't, already tried.

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h (hunk 2 above)
Comment (sonicdcer) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h:12]: It was a bad guess at the beginning i forgot to remove, thanks.

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c (hunk 1 above)
Comment (AngheloAlf) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:55]: What about with `% 16U`?

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 7]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
+}
+
+void DmTsg_Draw(Actor* thisx, PlayState* play2) {
+    PlayState* play = play2;
+    DmTsg* this = THIS;
+    s32 i;
+    u32 j;
+
+    if (this->unk_91E) {
+        for (i = 0, j = 0; i < 100; i++) {
```
Comment (isghj5) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:81]: sure looks like unk_91E is a boolean marking the actor safe to draw. `canDraw` or `isDrawable`?

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 8]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
 
-#endif
+void DmTsg_Init(Actor* thisx, PlayState* play) {
+    DmTsg* this = THIS;
+    s32 i;
 
-extern UNK_TYPE D_06011458;
+    if (gSaveContext.save.entrance == ENTRANCE(OPENING_DUNGEON, 0)) {
+        Actor_SetScale(&this->actor, 0.1f);
+        for (i = 0; i < 100; i++) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:37]:
```suggestion
        for (i = 0; i < DMTSG_UNK_SIZE; i++) {
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 9]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Tsg/DmTsg_Destroy.s")
+void DmTsg_Update(Actor* thisx, PlayState* play) {
+    DmTsg* this = THIS;
+    s32 cueChannel;
+    s16 i;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Dm_Tsg/DmTsg_Update.s")
+    for (i = 0; i < 100; i++) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:54]:
```suggestion
    for (i = 0; i < DMTSG_UNK_SIZE; i++) {
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c (hunk 7 above)
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:81]:
```suggestion
        for (i = 0, j = 0; i < DMTSG_UNK_SIZE; i++) {
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 10]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
+void DmTsg_Init(Actor* thisx, PlayState* play) {
+    DmTsg* this = THIS;
+    s32 i;
 
-extern UNK_TYPE D_06011458;
+    if (gSaveContext.save.entrance == ENTRANCE(OPENING_DUNGEON, 0)) {
+        Actor_SetScale(&this->actor, 0.1f);
+        for (i = 0; i < 100; i++) {
+            this->unk_78C[i] = (Rand_ZeroOne() * 65535.0f);
+            this->unk_856[i] = (s16)(Rand_ZeroOne() * 200.0f) + 100;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:39]:
```suggestion
            this->unk_856[i] = TRUNCF_BINANG(Rand_ZeroOne() * 200.0f) + 100;
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h [hunk 11]
```diff
@@ -5,9 +5,20 @@
...
 struct DmTsg;
 
+typedef void (*DmTsgActionFunc)(struct DmTsg*, PlayState*);
+
+#define DMTSG_UNK_SIZE 100
+
 typedef struct DmTsg {
     /* 0x000 */ Actor actor;
-    /* 0x144 */ char unk_144[0x7DC];
+    /* 0x144 */ UNK_TYPE4 unk_144;
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h:14]:
```suggestion
    /* 0x144 */ UNK_TYPE1 unk_144[0x4];
```

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h [hunk 12]
```diff
@@ -5,9 +5,20 @@
 
 struct DmTsg;
 
+typedef void (*DmTsgActionFunc)(struct DmTsg*, PlayState*);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.h:8]: Please remove this since it is unused

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c (hunk 1 above)
Comment (sonicdcer) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:55]: I think & 15 is more clear than % 16U

Comment (AngheloAlf) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c]: Why would doing a bitwise operation with a dec number be clearer than an hex number?
Also, in this same scope the code is operating `i` with 16, so using 16 is more consistent

Comment (sonicdcer) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c]: Yeah i didn't thought about that, i'll change it to % 16U although i don't see a dev writing that U in my mind for some reason.

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c [hunk 13]
```diff
@@ -28,14 +28,66 @@ ActorInit Dm_Tsg_InitVars = {
...
-#endif
+void DmTsg_Init(Actor* thisx, PlayState* play) {
+    DmTsg* this = THIS;
+    s32 i;
 
-extern UNK_TYPE D_06011458;
+    if (gSaveContext.save.entrance == ENTRANCE(OPENING_DUNGEON, 0)) {
+        Actor_SetScale(&this->actor, 0.1f);
+        for (i = 0; i < DMTSG_UNK_SIZE; i++) {
+            this->unk_78C[i] = (Rand_ZeroOne() * 65535.0f);
```
Comment (hensldm) [inline src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:38]:
```suggestion
            this->unk_78C[i] = Rand_ZeroOne() * 65535.0f;
```
Small nit

**Review by hensldm — APPROVED** [review]
1 tiny nit

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c (hunk 13 above)
Comment (engineer124) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c:38]: Since `unk_78C` is a rotation, `65535.0f` can be in hex too

**Review by engineer124 — APPROVED** [review]

File: src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c (hunk 1 above)
Comment (AngheloAlf) [reply src/overlays/actors/ovl_Dm_Tsg/z_dm_tsg.c]: Take in mind this can be the result of a macro, a `sizeof` or something like that

**Review by AngheloAlf — APPROVED** [review]

**Pull request merged**